File: src/rules/engine.js
import { evaluateCoreRules } from './core.js';
import { evaluateReadinessRules } from './readiness.js';
import { selectRepresentative } from '../domain/normalize.js';
import { CORE_RULE_IDS, V2_READINESS_RULE_IDS } from './catalogs.js';

/**
 * Rule execution grain and aggregation (design section 4; flow step 4).
 *
 * Core rules are evaluated on EVERY raw row, then aggregated to alert identity:
 *
 * - `count` per rule is the number of rows that actually match.
 * - `distinct_count` per rule is the number of identities with at least one matching row.
 * - Findings stay attached only to the rows that matched. They are never projected onto
 *   other rows sharing an identity, because that would move a finding onto a date where
 *   nothing was wrong.
 *
 * LLM eligibility is decided at IDENTITY level after that row-level pass: any core
 * finding anywhere in the window withholds the whole identity. That deliberately accepts
 * losing a second, advisory judgment for the alert — a deterministic finding already
 * gives the team a concrete fix, whereas evaluating only the representative could miss a
 * row-specific R5 suppression or R7 timestamp failure.
 */

/**
 * @typedef {import('../domain/normalize.js').AlertRecord} AlertRecord
 * @typedef {import('./core.js').Finding} Finding
 */

/**
 * @typedef {object} EvaluatedRow
 * @property {AlertRecord} row
 * @property {Finding[]} coreFindings
 * @property {Finding[]} readinessFindings
 */

/**
 * @typedef {object} EvaluatedIdentity
 * @property {string} identity
 * @property {'v1'|'v2'} schema
 * @property {string} application
 * @property {string} keyField
 * @property {AlertRecord} representative
 * @property {EvaluatedRow[]} rows
 * @property {string[]} coreRuleIds Distinct core rule ids matched anywhere in the window.
 * @property {string[]} readinessRuleIds Distinct readiness rule ids on the representative.
 * @property {boolean} hasCoreFinding
 * @property {boolean} llmEligible
 * @property {Set<string>} presentDates UTC dates this identity appears on.
 */

/**
 * @typedef {object} RuleBucketCount
 * @property {string} snapshotDate
 * @property {string} ruleId
 * @property {number} count Matching raw rows in this bucket.
 * @property {number} distinctCount Identities with at least one matching row in this bucket.
 */

/**
 * Order rule ids R1, R2, ... R10 numerically rather than lexically, so R10 does not sort
 * between R1 and R2 in reports.
 * @param {string} a
 * @param {string} b
 * @returns {number}
 */
export function compareRuleIds(a, b) {
  const na = Number(a.slice(1));
  const nb = Number(b.slice(1));
  if (Number.isFinite(na) && Number.isFinite(nb) && a[0] === b[0]) return na - nb;
  return a < b ? -1 : a > b ? 1 : 0;
}

/**
 * Evaluate every row of one schema and aggregate to identities.
 *
 * @param {AlertRecord[]} rows Rows of a single schema.
 * @returns {{rows: EvaluatedRow[], identities: Map<string, EvaluatedIdentity>}}
 */
export function evaluateRows(rows) {
  /** @type {EvaluatedRow[]} */
  const evaluated = rows.map((row) => ({
    row,
    coreFindings: evaluateCoreRules(row),
    readinessFindings: evaluateReadinessRules(row),
  }));

  /** @type {Map<string, EvaluatedRow[]>} */
  const grouped = new Map();
  for (const e of evaluated) {
    const bucket = grouped.get(e.row.identity);
    if (bucket) bucket.push(e);
    else grouped.set(e.row.identity, [e]);
  }

  /** @type {Map<string, EvaluatedIdentity>} */
  const identities = new Map();
  for (const [identity, group] of grouped) {
    const representative = selectRepresentative(group.map((e) => e.row));

    /** @type {Set<string>} */
    const coreRuleIds = new Set();
    for (const e of group) for (const f of e.coreFindings) coreRuleIds.add(f.ruleId);

    // Readiness is a property of the identity's current state, so it is read off the
    // representative rather than unioned over history: an alert enriched yesterday is
    // ready today, and a gap it used to have is not a gap now.
    const representativeEval = group.find((e) => e.row === representative);
    const readinessRuleIds = (representativeEval?.readinessFindings || []).map((f) => f.ruleId);

    /** @type {Set<string>} */
    const presentDates = new Set(group.map((e) => e.row.snapshotDate));

    const hasCoreFinding = coreRuleIds.size > 0;
    identities.set(identity, {
      identity,
      schema: representative.schema,
      application: representative.application,
      keyField: representative.keyField,
      representative,
      rows: group,
      coreRuleIds: [...coreRuleIds].sort(compareRuleIds),
      readinessRuleIds: readinessRuleIds.sort(compareRuleIds),
      hasCoreFinding,
      llmEligible: !hasCoreFinding,
      presentDates,
    });
  }

  return { rows: evaluated, identities };
}

/**
 * Attach externally-derived findings (R5 from the suppression evaluator) to already
 * evaluated rows, then recompute the identity-level aggregates that depend on them.
 *
 * R5 arrives late because suppression needs the team's panels and the complete owned row
 * set, which a single-row rule cannot see. It is a core finding like any other, so it
 * must be able to withhold an identity from the LLM.
 *
 * @param {{rows: EvaluatedRow[], identities: Map<string, EvaluatedIdentity>}} evaluation
 * @param {Map<AlertRecord, Finding>} findingsByRow
 * @returns {void}
 */
export function attachRowFindings(evaluation, findingsByRow) {
  if (findingsByRow.size === 0) return;

  for (const e of evaluation.rows) {
    const finding = findingsByRow.get(e.row);
    if (!finding) continue;
    if (finding.set === 'core') e.coreFindings.push(finding);
    else e.readinessFindings.push(finding);
  }

  for (const identity of evaluation.identities.values()) {
    /** @type {Set<string>} */
    const coreRuleIds = new Set();
    for (const e of identity.rows) for (const f of e.coreFindings) coreRuleIds.add(f.ruleId);
    identity.coreRuleIds = [...coreRuleIds].sort(compareRuleIds);
    identity.hasCoreFinding = coreRuleIds.size > 0;
    identity.llmEligible = !identity.hasCoreFinding;
  }
}

/**
 * Per-rule daily counts.
 *
 * A deterministic match is attributed to the UTC bucket containing the raw row that
 * matched. `distinctCount` counts an identity once in each bucket where at least one of
 * its rows matched — not once per window, and not on dates where it did not match.
 *
 * Readiness rules R8-R10 are counted here too, so the report can show the gap breakdown,
 * but they are excluded from `flagged_by_rule` below.
 *
 * @param {EvaluatedRow[]} evaluatedRows
 * @param {string[]} snapshotDates Every bucket date in the window, ascending.
 * @returns {RuleBucketCount[]}
 */
export function computeDailyRuleCounts(evaluatedRows, snapshotDates) {
  /** @type {Map<string, {count: number, identities: Set<string>}>} keyed by date and rule */
  const acc = new Map();

  for (const e of evaluatedRows) {
    const date = e.row.snapshotDate;
    for (const finding of [...e.coreFindings, ...e.readinessFindings]) {
      const key = `${date}|${finding.ruleId}`;
      let entry = acc.get(key);
      if (!entry) {
        entry = { count: 0, identities: new Set() };
        acc.set(key, entry);
      }
      entry.count += 1;
      entry.identities.add(e.row.identity);
    }
  }

  /** @type {RuleBucketCount[]} */
  const out = [];
  for (const snapshotDate of snapshotDates) {
    for (const ruleId of [...CORE_RULE_IDS, ...V2_READINESS_RULE_IDS]) {
      const entry = acc.get(`${snapshotDate}|${ruleId}`);
      if (!entry) continue;
      out.push({
        snapshotDate,
        ruleId,
        count: entry.count,
        distinctCount: entry.identities.size,
      });
    }
  }
  return out;
}

/**
 * `flagged_by_rule` and `flagged_by_rule_distinct` per bucket.
 *
 * The union of rows with at least one CORE finding, and the count of identities with at
 * least one core match in that bucket. Readiness gaps are excluded by construction.
 *
 * @param {EvaluatedRow[]} evaluatedRows
 * @param {string[]} snapshotDates
 * @returns {Map<string, {flaggedByRule: number, flaggedByRuleDistinct: number}>}
 */
export function computeDailyFlagged(evaluatedRows, snapshotDates) {
  /** @type {Map<string, {rows: number, identities: Set<string>}>} */
  const acc = new Map();
  for (const date of snapshotDates) acc.set(date, { rows: 0, identities: new Set() });

  for (const e of evaluatedRows) {
    if (e.coreFindings.length === 0) continue;
    const entry = acc.get(e.row.snapshotDate);
    if (!entry) continue;
    entry.rows += 1;
    entry.identities.add(e.row.identity);
  }

  /** @type {Map<string, {flaggedByRule: number, flaggedByRuleDistinct: number}>} */
  const out = new Map();
  for (const [date, entry] of acc) {
    out.set(date, { flaggedByRule: entry.rows, flaggedByRuleDistinct: entry.identities.size });
  }
  return out;
}

/**
 * Count of v2 identities carrying at least one readiness gap on their representative.
 * @param {Iterable<EvaluatedIdentity>} identities
 * @returns {number}
 */
export function countPhase2GapIdentities(identities) {
  let n = 0;
  for (const identity of identities) {
    if (identity.schema === 'v2' && identity.readinessRuleIds.length > 0) n += 1;
  }
  return n;
}

File: src/alerts_bi/rules/engine.py
"""Rule execution grain and aggregation (design section 4; flow step 4).

Core rules are evaluated on EVERY raw row, then aggregated to alert identity:

- ``count`` per rule is the number of rows that actually match.
- ``distinct_count`` per rule is the number of identities with at least one matching row.
- Findings stay attached only to the rows that matched. They are never projected onto
  other rows sharing an identity, because that would move a finding onto a date where
  nothing was wrong.

LLM eligibility is decided at IDENTITY level after that row-level pass: any core finding
anywhere in the window withholds the whole identity. That deliberately accepts losing a
second, advisory judgment for the alert - a deterministic finding already gives the team a
concrete fix, whereas evaluating only the representative could miss a row-specific R5
suppression or R7 timestamp failure.
"""

from __future__ import annotations

from collections.abc import Mapping, Sequence
from dataclasses import dataclass, field

from alerts_bi.domain.normalize import AlertRecord, select_representative
from alerts_bi.rules.catalogs import CORE_RULE_IDS, V2_READINESS_RULE_IDS
from alerts_bi.rules.core import Finding, evaluate_core_rules
from alerts_bi.rules.readiness import evaluate_readiness_rules

__all__ = [
    "EvaluatedIdentity",
    "EvaluatedRow",
    "Evaluation",
    "RuleBucketCount",
    "attach_row_findings",
    "compare_rule_ids",
    "compute_daily_flagged",
    "compute_daily_rule_counts",
    "count_phase2_gap_identities",
    "evaluate_rows",
]


def compare_rule_ids(rule_id: str) -> tuple[str, int]:
    """Sort key ordering R1, R2, ... R10 numerically rather than lexically.

    Without this R10 sorts between R1 and R2 in reports.
    """
    prefix = rule_id[0]
    try:
        return (prefix, int(rule_id[1:]))
    except ValueError:
        return (prefix, 0)


@dataclass(slots=True)
class EvaluatedRow:
    row: AlertRecord
    core_findings: list[Finding] = field(default_factory=list)
    readiness_findings: list[Finding] = field(default_factory=list)


@dataclass(slots=True)
class EvaluatedIdentity:
    identity: str
    schema: str
    application: str
    key_field: str
    representative: AlertRecord
    rows: list[EvaluatedRow]
    core_rule_ids: list[str]
    """Distinct core rule ids matched anywhere in the window."""
    readiness_rule_ids: list[str]
    """Distinct readiness rule ids on the representative."""
    has_core_finding: bool
    llm_eligible: bool
    present_dates: set[str]
    """UTC dates this identity appears on."""


@dataclass(slots=True)
class Evaluation:
    rows: list[EvaluatedRow]
    identities: dict[str, EvaluatedIdentity]


@dataclass(frozen=True, slots=True)
class RuleBucketCount:
    snapshot_date: str
    rule_id: str
    match_count: int
    """Matching raw rows in this bucket. Named to match the SQL column, and because a
    NamedTuple field called ``count`` would shadow ``tuple.count``."""
    distinct_count: int
    """Identities with at least one matching row in this bucket."""


def evaluate_rows(rows: Sequence[AlertRecord]) -> Evaluation:
    """Evaluate every row of one schema and aggregate to identities."""
    evaluated = [
        EvaluatedRow(
            row=row,
            core_findings=evaluate_core_rules(row),
            readiness_findings=evaluate_readiness_rules(row),
        )
        for row in rows
    ]

    grouped: dict[str, list[EvaluatedRow]] = {}
    for item in evaluated:
        grouped.setdefault(item.row.identity, []).append(item)

    identities: dict[str, EvaluatedIdentity] = {}
    for identity, group in grouped.items():
        representative = select_representative([item.row for item in group])

        core_rule_ids = {finding.rule_id for item in group for finding in item.core_findings}

        # Readiness is a property of the identity's CURRENT state, so it is read off the
        # representative rather than unioned over history: an alert enriched yesterday is
        # ready today, and a gap it used to have is not a gap now (design section 7.6).
        representative_eval = next(item for item in group if item.row is representative)
        readiness_rule_ids = [f.rule_id for f in representative_eval.readiness_findings]

        has_core_finding = bool(core_rule_ids)
        identities[identity] = EvaluatedIdentity(
            identity=identity,
            schema=representative.schema,
            application=representative.application,
            key_field=representative.key_field,
            representative=representative,
            rows=group,
            core_rule_ids=sorted(core_rule_ids, key=compare_rule_ids),
            readiness_rule_ids=sorted(readiness_rule_ids, key=compare_rule_ids),
            has_core_finding=has_core_finding,
            llm_eligible=not has_core_finding,
            present_dates={item.row.snapshot_date for item in group},
        )

    return Evaluation(rows=evaluated, identities=identities)


def attach_row_findings(evaluation: Evaluation, findings_by_row: Mapping[int, Finding]) -> None:
    """Attach externally-derived findings (R5) to evaluated rows, then recompute.

    R5 arrives late because suppression needs the team's panels and the complete owned row
    set, which a single-row rule cannot see. It is a core finding like any other, so it
    must be able to withhold an identity from the LLM.

    Keyed by ``id(row)`` because :class:`AlertRecord` is frozen but not hashable by value:
    two rows of a re-fired alert can be field-identical, and they must stay distinguishable.
    """
    if not findings_by_row:
        return

    for item in evaluation.rows:
        finding = findings_by_row.get(id(item.row))
        if finding is None:
            continue
        if finding.set == "core":
            item.core_findings.append(finding)
        else:
            item.readiness_findings.append(finding)

    for identity in evaluation.identities.values():
        core_rule_ids = {
            finding.rule_id for item in identity.rows for finding in item.core_findings
        }
        identity.core_rule_ids = sorted(core_rule_ids, key=compare_rule_ids)
        identity.has_core_finding = bool(core_rule_ids)
        identity.llm_eligible = not identity.has_core_finding


def compute_daily_rule_counts(
    evaluated_rows: Sequence[EvaluatedRow], snapshot_dates: Sequence[str]
) -> list[RuleBucketCount]:
    """Per-rule daily counts.

    A deterministic match is attributed to the UTC bucket containing the raw row that
    matched. ``distinct_count`` counts an identity once in each bucket where at least one
    of its rows matched - not once per window, and not on dates where it did not match.

    Readiness rules R8-R10 are counted here too, so the report can show the gap breakdown,
    but they are excluded from ``flagged_by_rule`` below.
    """
    counts: dict[tuple[str, str], int] = {}
    identities: dict[tuple[str, str], set[str]] = {}

    for item in evaluated_rows:
        date = item.row.snapshot_date
        for finding in (*item.core_findings, *item.readiness_findings):
            key = (date, finding.rule_id)
            counts[key] = counts.get(key, 0) + 1
            identities.setdefault(key, set()).add(item.row.identity)

    out: list[RuleBucketCount] = []
    for snapshot_date in snapshot_dates:
        for rule_id in (*CORE_RULE_IDS, *V2_READINESS_RULE_IDS):
            key = (snapshot_date, rule_id)
            if key not in counts:
                continue
            out.append(
                RuleBucketCount(
                    snapshot_date=snapshot_date,
                    rule_id=rule_id,
                    match_count=counts[key],
                    distinct_count=len(identities[key]),
                )
            )
    return out


def compute_daily_flagged(
    evaluated_rows: Sequence[EvaluatedRow], snapshot_dates: Sequence[str]
) -> dict[str, tuple[int, int]]:
    """``flagged_by_rule`` and ``flagged_by_rule_distinct`` per bucket.

    The union of rows with at least one CORE finding, and the count of identities with at
    least one core match in that bucket. Readiness gaps are excluded by construction.

    Returns ``{snapshot_date: (flagged_by_rule, flagged_by_rule_distinct)}``.
    """
    rows_per_date = dict.fromkeys(snapshot_dates, 0)
    identities_per_date: dict[str, set[str]] = {date: set() for date in snapshot_dates}

    for item in evaluated_rows:
        if not item.core_findings:
            continue
        date = item.row.snapshot_date
        if date not in rows_per_date:
            continue
        rows_per_date[date] += 1
        identities_per_date[date].add(item.row.identity)

    return {date: (rows_per_date[date], len(identities_per_date[date])) for date in snapshot_dates}


def count_phase2_gap_identities(identities: Sequence[EvaluatedIdentity]) -> int:
    """Count of v2 identities carrying at least one readiness gap on their representative."""
    return sum(1 for i in identities if i.schema == "v2" and i.readiness_rule_ids)
